rpc/src/client/transport/websocket.rs

```diff
@@ -0,0 +1,701 @@
+impl SubscriptionClient for WebSocketSubscriptionClient {
+    async fn subscribe(&mut self, query: String) -> Result<Subscription> {
+        let (event_tx, event_rx) = unbounded();
+        let (result_tx, mut result_rx) = unbounded::<Result<()>>();
+        // NB: We assume here that the wrapper generates a unique ID for our
+        // subscription.
+        let req = request::Wrapper::new(subscribe::Request::new(query.clone()));
+        let id: SubscriptionId = req.id().clone().try_into()?;
+        self.send_cmd(WebSocketDriverCmd::Subscribe {
+            req,
+            event_tx,
+            result_tx,
+        })
+        .await?;
+        // Wait to make sure our subscription request went through
+        // successfully.
+        result_rx.recv().await.ok_or_else(|| {
+            Error::client_internal_error("failed to hear back from WebSocket driver".to_string())
+        })??;
+        Ok(Subscription::new(
+            id,
+            query,
+            event_rx,
+            self.terminate_tx.clone(),
+        ))
+    }
+}
+
+#[async_trait]
+impl ClosableClient for WebSocketSubscriptionClient {
+    /// Attempt to gracefully close the WebSocket connection.
+    async fn close(mut self) -> Result<()> {
+        self.cmd_tx.send(WebSocketDriverCmd::Close).await?;
+        self.driver_handle.await.map_err(|e| {
+            Error::client_internal_error(format!(
+                "failed while waiting for WebSocket driver task to terminate: {}",
+                e
+            ))
+        })?
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct GenericJSONResponse(serde_json::Value);
+
+impl Response for GenericJSONResponse {}
+
+#[derive(Debug)]
+struct WebSocketSubscriptionDriver {
+    stream: WebSocketStream<TokioAdapter<TcpStream>>,
+    router: SubscriptionRouter,
+    cmd_rx: ChannelRx<WebSocketDriverCmd>,
+    terminate_rx: ChannelRx<TerminateSubscription>,
+}
+
+impl WebSocketSubscriptionDriver {
+    fn new(
+        stream: WebSocketStream<TokioAdapter<TcpStream>>,
+        cmd_rx: ChannelRx<WebSocketDriverCmd>,
+        terminate_rx: ChannelRx<TerminateSubscription>,
+    ) -> Self {
+        Self {
+            stream,
+            router: SubscriptionRouter::default(),
+            cmd_rx,
+            terminate_rx,
+        }
+    }
+
+    async fn run(mut self) -> Result<()> {
+        // TODO(thane): Should this loop initiate a keepalive (ping) to the
+        //              server on a regular basis?
+        loop {
+            tokio::select! {
+                Some(res) = self.stream.next() => match res {
+                    Ok(msg) => self.handle_incoming_msg(msg).await?,
+                    Err(e) => return Err(
+                        Error::websocket_error(
+                            format!("failed to read from WebSocket connection: {}", e),
+                        ),
+                    ),
+                },
+                Some(cmd) = self.cmd_rx.recv() => match cmd {
```

Comment:
I find it a bit confusing that subscriptions come over the cmd_rx channel but the unsubscriptions come over a separate terminate channel. Why not give the Susbcription objects the cmd_tx and let them send Unsubscribe commands over that?